File: ios/Sources/PoltioSDK/PoltioSDK.swift
import Foundation

/// Main entry point for the Poltio iOS SDK.
public final class PoltioSDK {
    /// Shared singleton instance of `PoltioSDK`.
    public static let shared = PoltioSDK()

    private var _clientKey: String?
    private var _isInitialized: Bool = false
    private var _sdkId: String?
    private var _puid: String?
    private var _puidLoaded: Bool = false
    private var _currentViewRequestId: UInt64 = 0

    private var _activeWidgetTask: URLSessionDataTask?
    private let _widgetCache = PoltioWidgetCache()

    private let queue = DispatchQueue(label: "com.poltio.sdk", attributes: .concurrent)

    private static let sdkIdStorageKey = "com.poltio.sdk.sdk_id"
    private static let puidStorageKey = "com.poltio.sdk.puid"

    /// The active log level for the SDK.
    public static var logLevel: PoltioLogLevel {
        get { PoltioLogger.logLevel }
        set { PoltioLogger.logLevel = newValue }
    }

    /// Time-to-live in seconds for widget resolution in-memory cache (default: 300 seconds / 5 minutes).
    public static var cacheTTL: TimeInterval {
        get { shared.cacheTTL }
        set { shared.cacheTTL = newValue }
    }

    /// Instance property for cache TTL.
    public var cacheTTL: TimeInterval {
        get { _widgetCache.defaultTTL }
        set { _widgetCache.defaultTTL = newValue }
    }

    /// Maximum number of widget responses retained in the in-memory cache (default: 100).
    public static var cacheLimit: Int {
        get { shared.cacheLimit }
        set { shared.cacheLimit = newValue }
    }

    /// Instance property for cache item limit.
    public var cacheLimit: Int {
        get { _widgetCache.countLimit }
        set { _widgetCache.countLimit = newValue }
    }

    /// Clears the in-memory widget resolution cache.
    public static func clearCache() {
        shared.clearCache()
    }

    /// Instance method to clear the in-memory widget cache.
    public func clearCache() {
        _widgetCache.clear()
    }

    /// Internal access to the widget cache (for unit testing).
    var widgetCache: PoltioWidgetCache {
        _widgetCache
    }

    /// The client key configured for this SDK session.
    public var clientKey: String? {
        queue.sync { _clientKey }
    }

    /// Indicates whether the SDK has been configured.
    public var isInitialized: Bool {
        queue.sync { _isInitialized }
    }

    /// SDK-generated unique identifier (source of truth for device tracking).
    /// Generated automatically on first access and persisted in local storage.
    public var sdkId: String {
        if let existingId = queue.sync(execute: { _sdkId }) {
            return existingId
        }

        return queue.sync(flags: .barrier) {
            if let existingId = self._sdkId {
                return existingId
            }

            if let persistedId = UserDefaults.standard.string(forKey: PoltioSDK.sdkIdStorageKey), !persistedId.isEmpty {
                self._sdkId = persistedId
                return persistedId
            }

            let newId = UUID().uuidString
            UserDefaults.standard.set(newId, forKey: PoltioSDK.sdkIdStorageKey)
            self._sdkId = newId
            return newId
        }
    }

    /// Developer-provided optional user identifier (PUID).
    public var puid: String? {
        let (loaded, cached) = queue.sync { (self._puidLoaded, self._puid) }
        if loaded {
            return cached
        }

        return queue.sync(flags: .barrier) {
            if self._puidLoaded {
                return self._puid
            }

            let persisted = UserDefaults.standard.string(forKey: PoltioSDK.puidStorageKey)
            self._puid = persisted
            self._puidLoaded = true
            return persisted
        }
    }

    private var _apiClient: PoltioAPIClient?

    /// The API client instance used for backend requests.
    var apiClient: PoltioAPIClient {
        get {
            if let existing = queue.sync(execute: { _apiClient }) {
                return existing
            }
            return queue.sync(flags: .barrier) {
                if let existing = self._apiClient {
                    return existing
                }
                let client = PoltioAPIClient()
                self._apiClient = client
                return client
            }
        }
        set {
            queue.sync(flags: .barrier) {
                _apiClient = newValue
            }
        }
    }

    init() {}

    #if DEBUG
        /// Resets the SDK state. Used exclusively for unit testing to prevent state pollution.
        func reset() {
            queue.sync(flags: .barrier) {
                self._clientKey = nil
                self._isInitialized = false
                self._puid = nil
                self._puidLoaded = false
                self._sdkId = nil
                self._apiClient = nil
                self._activeWidgetTask?.cancel()
                self._activeWidgetTask = nil
                self._currentViewRequestId = 0
                self._widgetCache.clear()
                self._widgetCache.defaultTTL = 300.0
                self._widgetCache.countLimit = 100
                UserDefaults.standard.removeObject(forKey: PoltioSDK.sdkIdStorageKey)
                UserDefaults.standard.removeObject(forKey: PoltioSDK.puidStorageKey)
            }
            PoltioLogger.logLevel = .info
        }
    #endif

    // MARK: - Public Configuration API

    /// Configures the Poltio SDK with your publishable client key and optional log level.
    /// - Parameters:
    ///   - clientKey: Poltio client key (e.g. "poltio_test_pk...")
    ///   - logLevel: Verbosity level of console logging (defaults to `.info`).
    /// - Returns: The configured shared instance.
    @discardableResult
    public static func configure(
        clientKey: String,
        logLevel: PoltioLogLevel = .info
    ) -> PoltioSDK {
        shared.configure(clientKey: clientKey, logLevel: logLevel)
    }

    /// Instance method to configure the SDK.
    @discardableResult
    func configure(
        clientKey: String,
        logLevel: PoltioLogLevel = .info
    ) -> PoltioSDK {
        let trimmedKey = clientKey.trimmingCharacters(in: .whitespacesAndNewlines)
        PoltioLogger.logLevel = logLevel

        guard !trimmedKey.isEmpty else {
            PoltioLogger.error("Client key cannot be empty.")
            return self
        }

        queue.sync(flags: .barrier) {
            self._clientKey = trimmedKey
            self._isInitialized = true
        }

        PoltioLogger.info("Configured successfully (SDK ID: \(sdkId)).")
        return self
    }

    // MARK: - Public User Identification API

    /// Identifies the user with an optional developer-provided user identifier (PUID).
    /// - Parameter puid: Optional developer-provided unique user identifier (pass nil or empty string to clear).
    public static func identify(puid: String?) {
        shared.identify(puid: puid)
    }

    /// Instance method to set or clear developer-provided PUID.
    func identify(puid: String?) {
        let trimmedPuid = puid?.trimmingCharacters(in: .whitespacesAndNewlines)

        queue.sync(flags: .barrier) {
            if let validPuid = trimmedPuid, !validPuid.isEmpty {
                self._puid = validPuid
                UserDefaults.standard.set(validPuid, forKey: PoltioSDK.puidStorageKey)
                PoltioLogger.info("Identified user with PUID: '\(validPuid)'.")
            } else {
                self._puid = nil
                UserDefaults.standard.removeObject(forKey: PoltioSDK.puidStorageKey)
                PoltioLogger.info("Cleared PUID.")
            }
            self._puidLoaded = true
        }
    }

    // MARK: - Public Event Tracking API

    /// Tracks an in-app event with optional parameters.
    /// Automatically attaches `sdk_id` and `puid` (when available) to the event parameters.
    /// For `view` events, triggers backend widget resolution via `/sdk/mobile/v1/widget`.
    /// - Parameters:
    ///   - event: The event name (e.g. "view", "TrackConversion")
    ///   - params: Dictionary of event properties (e.g. ["url": "https://www.poltio.com/pdp"])
    public static func track(event: String, params: [String: Any]? = nil) {
        shared.track(event: event, params: params)
    }

    /// Instance method to track an in-app event.
    func track(event: String, params: [String: Any]? = nil) {
        let trimmedEvent = event.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedEvent.isEmpty else {
            PoltioLogger.error("Event name cannot be empty.")
            return
        }

        guard isInitialized, let key = clientKey else {
            PoltioLogger.warning("track called before configuration. Call PoltioSDK.configure(clientKey:) first.")
            return
        }

        var enrichedParams: [String: Any] = params ?? [:]
        let currentSdkId = sdkId
        enrichedParams["sdk_id"] = currentSdkId
        if let currentPuid = puid {
            enrichedParams["puid"] = currentPuid
        }

        PoltioLogger.info("Event tracked: '\(trimmedEvent)', params: \(enrichedParams)")

        if isViewEvent(trimmedEvent) {
            let rawUrl: String = if let value = params?["url"] ?? params?["screen"] ?? params?["page"] {
                (value as? String) ?? (value as? URL)?.absoluteString ?? ""
            } else {
                ""
            }
            let targetURL = PoltioSDK.sanitizeOrFormatURL(rawUrl)
            let activePuid = puid

            // Check in-memory cache first
            if let cachedResult = _widgetCache.get(for: targetURL) {
                queue.sync(flags: .barrier) {
                    self._activeWidgetTask?.cancel()
                    self._activeWidgetTask = nil
                    self._currentViewRequestId &+= 1
                }

                switch cachedResult {
                case let .widget(widgetResponse):
                    PoltioLogger.debug("Using cached widget response for '\(targetURL)' (public_id: \(widgetResponse.publicId)).")
                    #if canImport(UIKit)
                        PoltioOverlayManager.shared.showTrigger(widget: widgetResponse, puid: activePuid)
                    #endif
                case .noWidget:
                    PoltioLogger.debug("Using cached negative widget resolution for '\(targetURL)' (no widget).")
                    #if canImport(UIKit)
                        PoltioOverlayManager.shared.hideTrigger()
                    #endif
                }
                return
            }

            // Cancel any in-flight widget resolution task for previous screen
            queue.sync(flags: .barrier) {
                self._activeWidgetTask?.cancel()
                self._activeWidgetTask = nil
                self._currentViewRequestId &+= 1
            }

            let thisRequestId: UInt64 = queue.sync { self._currentViewRequestId }

            let task = apiClient.resolveMobileWidget(
                clientKey: key,
                deviceId: currentSdkId,
                targetURL: targetURL
            ) { [weak self] result in
                guard let self else { return }

                let isLatest: Bool = queue.sync {
                    thisRequestId == self._currentViewRequestId
                }

                guard isLatest else {
                    PoltioLogger.debug("Ignoring outdated widget resolution result for '\(targetURL)' (newer screen was already requested).")
                    return
                }

                switch result {
                case let .success(widgetResponse):
                    _widgetCache.set(result: .widget(widgetResponse), for: targetURL)
                    #if canImport(UIKit)
                        PoltioOverlayManager.shared.showTrigger(widget: widgetResponse, puid: activePuid)
                    #endif
                case let .failure(error):
                    #if canImport(UIKit)
                        PoltioOverlayManager.shared.hideTrigger()
                    #endif

                    if (error as? URLError)?.code == .resourceUnavailable {
                        _widgetCache.set(result: .noWidget, for: targetURL)
                    }

                    if (error as? URLError)?.code != .cancelled, (error as NSError).code != NSURLErrorCancelled {
                        PoltioLogger.warning("Widget resolution skipped/failed for '\(targetURL)': \(error.localizedDescription)")
                    }
                }
            }

            queue.sync(flags: .barrier) {
                if thisRequestId == self._currentViewRequestId {
                    self._activeWidgetTask = task
                } else {
                    // A newer screen already superseded this request before it could be stored;
                    // cancel it explicitly instead of letting it run to completion unused.
                    task?.cancel()
                }
            }
        }
    }

    // MARK: - Public Widget Event Bridge

    /// Optional callback invoked when the widget WebView emits a bridge event (e.g. "close", "complete",
    /// "leadSubmit"). The widget page communicates via
    /// `window.webkit.messageHandlers.poltioNative.postMessage({ event: "...", data: { ... } })`.
    /// Delivered on the main thread. Set this once, e.g. alongside `configure(clientKey:)`.
    public static var onWidgetEvent: ((_ event: String, _ data: [String: Any]?) -> Void)?

    // MARK: - Internal Helpers

    /// Checks whether an event name corresponds to a view event.
    func isViewEvent(_ eventName: String) -> Bool {
        let lower = eventName.lowercased()
        return lower == "view" || lower == "viewcontent" || lower == "view_content"
    }

    /// Sanitizes or formats a raw URL string to guarantee it contains a scheme and host required by the API.
    static func sanitizeOrFormatURL(_ rawInput: String) -> String {
        let trimmed = rawInput.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return "https://app.poltio.com/default"
        }

        if trimmed.contains("://") {
            // Decide whether encoding is needed by inspecting the actual characters present, not by
            // trusting `URL(string:)` to reject invalid input — Foundation's URL parser is lenient enough
            // to happily parse raw, unescaped spaces, so a "does it parse?" check can't distinguish
            // "already valid" from "needs encoding".
            let needsEncoding = trimmed.rangeOfCharacter(from: CharacterSet.poltioURLAllowed.inverted) != nil
            guard needsEncoding else {
                // Already well-formed (including already percent-encoded). Re-encoding here would
                // double-encode existing '%XX' sequences (turning "%20" into "%2520") and corrupt the
                // target URL sent to the widget resolution endpoint.
                return trimmed
            }

            // Encode only the characters that actually need it, leaving existing URL syntax
            // (scheme/path/query/fragment delimiters and already-escaped '%XX' sequences) untouched.
            if let encoded = trimmed.addingPercentEncoding(withAllowedCharacters: .poltioURLAllowed),
               let parsed = URL(string: encoded), parsed.scheme != nil, parsed.host != nil
            {
                return encoded
            }
            return trimmed
        }

        if let parsed = URL(string: trimmed), parsed.scheme != nil, parsed.host != nil {
            return trimmed
        }

        let cleanPath = trimmed.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return "https://app.poltio.com/\(cleanPath)"
    }
}

private extension CharacterSet {
    /// Broader than `.urlQueryAllowed`: covers the full set of characters that are valid, unescaped,
    /// anywhere in a URL (scheme, path, query, or fragment), so a fallback encoding pass doesn't
    /// escape delimiters like ':', '/', '#', '?' or already-valid '%' escape sequences.
    static let poltioURLAllowed: CharacterSet = {
        var set = CharacterSet.alphanumerics
        set.insert(charactersIn: "-._~:/?#[]@!$&'()*+,;=%")
        return set
    }()
}

File: android/poltio-sdk/src/main/java/com/poltio/sdk/PoltioSDK.kt
package com.poltio.sdk

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import androidx.annotation.VisibleForTesting
import com.poltio.sdk.ui.PoltioOverlayManager
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong

/**
 * Main entry point for the Poltio Android SDK. A plain Kotlin `object` since the SDK is
 * process-wide singleton state — there is no `PoltioSDK.shared` indirection like on iOS.
 */
object PoltioSDK {
    private const val PREFS_NAME = "com.poltio.sdk.prefs"
    private const val SDK_ID_KEY = "sdk_id"
    private const val PUID_KEY = "puid"

    private val lock = Any()
    private var applicationContext: Context? = null
    private var _clientKey: String? = null
    private var _isInitialized: Boolean = false
    private var _sdkId: String? = null
    private var _puid: String? = null
    private var _puidLoaded: Boolean = false
    private var _apiClient: PoltioAPIClient? = null

    private val widgetCache = PoltioWidgetCache()
    private val currentViewRequestId = AtomicLong(0)
    private var activeCall: PoltioAPIClient.Call? = null

    /** The active log level for the SDK. */
    var logLevel: PoltioLogLevel
        get() = PoltioLogger.logLevel
        set(value) { PoltioLogger.logLevel = value }

    /** Time-to-live in seconds for widget resolution in-memory cache (default: 300 seconds / 5 minutes). */
    var cacheTTL: Double
        get() = widgetCache.defaultTTL
        set(value) { widgetCache.defaultTTL = value }

    /** Maximum number of widget responses retained in the in-memory cache (default: 100). */
    var cacheLimit: Int
        get() = widgetCache.countLimit
        set(value) { widgetCache.countLimit = value }

    /** Clears the in-memory widget resolution cache. */
    fun clearCache() = widgetCache.clear()

    /** Internal access to the widget cache (for unit testing). */
    @VisibleForTesting
    internal fun widgetCacheForTesting(): PoltioWidgetCache = widgetCache

    /** The client key configured for this SDK session. */
    val clientKey: String? get() = synchronized(lock) { _clientKey }

    /** Indicates whether the SDK has been configured. */
    val isInitialized: Boolean get() = synchronized(lock) { _isInitialized }

    /**
     * SDK-generated unique identifier (source of truth for device tracking). Generated
     * automatically on first access and persisted in `SharedPreferences`. Requires [configure] to
     * have run at least once (to have a `Context` to persist against); before that, a fresh,
     * non-persisted id is minted on every call.
     */
    val sdkId: String
        get() {
            synchronized(lock) {
                _sdkId?.let { return it }

                val prefs = applicationContext?.let(::prefsFor)
                if (prefs != null) {
                    val persisted = prefs.getString(SDK_ID_KEY, null)
                    if (!persisted.isNullOrEmpty()) {
                        _sdkId = persisted
                        return persisted
                    }
                }

                val newId = UUID.randomUUID().toString()
                prefs?.edit()?.putString(SDK_ID_KEY, newId)?.apply()
                if (prefs == null) {
                    PoltioLogger.warning { "sdkId requested before configure(context, ...) — using a non-persisted id for this call." }
                } else {
                    _sdkId = newId
                }
                return newId
            }
        }

    /** Developer-provided optional user identifier (PUID). */
    val puid: String?
        get() {
            synchronized(lock) {
                if (_puidLoaded) return _puid
                val persisted = applicationContext?.let(::prefsFor)?.getString(PUID_KEY, null)
                _puid = persisted
                _puidLoaded = true
                return persisted
            }
        }

    /** The API client instance used for backend requests. */
    internal var apiClient: PoltioAPIClient
        get() = synchronized(lock) {
            _apiClient ?: PoltioAPIClient().also { _apiClient = it }
        }
        set(value) = synchronized(lock) { _apiClient = value }

    private fun prefsFor(context: Context): SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /** Resets the SDK state. Used exclusively for unit testing to prevent state pollution. */
    @VisibleForTesting
    internal fun reset() {
        synchronized(lock) {
            activeCall?.cancel()
            activeCall = null
            currentViewRequestId.set(0)
            applicationContext?.let { prefsFor(it).edit().clear().apply() }
            applicationContext = null
            _clientKey = null
            _isInitialized = false
            _puid = null
            _puidLoaded = false
            _sdkId = null
            _apiClient = null
            widgetCache.clear()
            widgetCache.defaultTTL = 300.0
            widgetCache.countLimit = 100
        }
        PoltioLogger.logLevel = PoltioLogLevel.INFO
    }

    // MARK: - Public Configuration API

    /**
     * Configures the Poltio SDK with your publishable client key and optional log level.
     *
     * @param context Any `Context`; only its `applicationContext` is retained, so passing an
     * `Activity` is safe and will not leak it.
     * @param clientKey Poltio client key (e.g. "poltio_test_pk...")
     * @param useStage Forces the stage (`true`) or production (`false`) API endpoint. Defaults to
     * `null`, which auto-detects the environment from the app's `android:debuggable` flag: debug
     * builds resolve to stage, release builds resolve to production. Pass an explicit value to
     * override this detection.
     * @param logLevel Verbosity level of console logging (defaults to `.INFO`).
     */
    fun configure(
        context: Context,
        clientKey: String,
        useStage: Boolean? = null,
        logLevel: PoltioLogLevel = PoltioLogLevel.INFO,
    ) {
        val trimmedKey = clientKey.trim()
        PoltioLogger.logLevel = logLevel

        if (trimmedKey.isEmpty()) {
            PoltioLogger.error { "Client key cannot be empty." }
            return
        }

        val appContext = context.applicationContext
        val environment = PoltioEnvironment.resolve(appContext, useStage)

        synchronized(lock) {
            applicationContext = appContext
            _clientKey = trimmedKey
            _isInitialized = true
            _apiClient = PoltioAPIClient(baseURL = environment.baseURL)
        }

        (appContext as? Application)?.let { PoltioOverlayManager.attach(it) }
            ?: PoltioLogger.warning { "configure() was not given an Application context — floating triggers cannot be attached to host activities." }

        if (environment == PoltioEnvironment.STAGE) {
            PoltioLogger.info {
                "Using STAGE API endpoint (${environment.baseURL}). Pass useStage = false to configure() once you're ready to point at production."
            }
        }

        PoltioLogger.info { "Configured successfully (SDK ID: $sdkId)." }
    }

    // MARK: - Public User Identification API

    /** Identifies the user with an optional developer-provided user identifier (PUID). */
    fun identify(puid: String?) {
        val trimmedPuid = puid?.trim()

        synchronized(lock) {
            val prefs = applicationContext?.let(::prefsFor)
            if (!trimmedPuid.isNullOrEmpty()) {
                _puid = trimmedPuid
                prefs?.edit()?.putString(PUID_KEY, trimmedPuid)?.apply()
                PoltioLogger.info { "Identified user with PUID: '$trimmedPuid'." }
            } else {
                _puid = null
                prefs?.edit()?.remove(PUID_KEY)?.apply()
                PoltioLogger.info { "Cleared PUID." }
            }
            _puidLoaded = true
        }
    }

    // MARK: - Public Event Tracking API

    /**
     * Tracks an in-app event with optional parameters. Automatically attaches `sdk_id` and `puid`
     * (when available) to the event parameters. For `view` events, triggers backend widget
     * resolution via `/sdk/mobile/v1/widget`.
     *
     * @param event The event name (e.g. "view", "TrackConversion")
     * @param params Map of event properties (e.g. `mapOf("url" to "https://www.poltio.com/pdp")`)
     */
    fun track(event: String, params: Map<String, Any?>? = null) {
        val trimmedEvent = event.trim()
        if (trimmedEvent.isEmpty()) {
            PoltioLogger.error { "Event name cannot be empty." }
            return
        }

        val key = clientKey
        if (!isInitialized || key == null) {
            PoltioLogger.warning { "track() called before configuration. Call PoltioSDK.configure(context, clientKey) first." }
            return
        }

        val enrichedParams = LinkedHashMap<String, Any?>(params ?: emptyMap())
        val currentSdkId = sdkId
        enrichedParams["sdk_id"] = currentSdkId
        puid?.let { enrichedParams["puid"] = it }

        PoltioLogger.info { "Event tracked: '$trimmedEvent', params: $enrichedParams" }

        if (isViewEvent(trimmedEvent)) {
            handleViewEvent(key, currentSdkId, params)
        }
    }

    private fun handleViewEvent(clientKey: String, deviceId: String, params: Map<String, Any?>?) {
        val rawUrl = (params?.get("url") ?: params?.get("screen") ?: params?.get("page"))?.toString() ?: ""
        val targetURL = sanitizeOrFormatURL(rawUrl)
        val activePuid = puid

        val cached = widgetCache.get(targetURL)
        if (cached != null) {
            synchronized(lock) {
                activeCall?.cancel()
                activeCall = null
                currentViewRequestId.incrementAndGet()
            }
            when (cached) {
                is CachedWidgetResult.Widget -> {
                    PoltioLogger.debug { "Using cached widget response for '$targetURL' (public_id: ${cached.response.publicId})." }
                    PoltioOverlayManager.showTrigger(cached.response, activePuid)
                }
                CachedWidgetResult.NoWidget -> {
                    PoltioLogger.debug { "Using cached negative widget resolution for '$targetURL' (no widget)." }
                    PoltioOverlayManager.hideTrigger()
                }
            }
            return
        }

        synchronized(lock) {
            activeCall?.cancel()
            activeCall = null
            currentViewRequestId.incrementAndGet()
        }
        val thisRequestId = currentViewRequestId.get()

        val call = apiClient.resolveMobileWidget(clientKey, deviceId, targetURL) { result ->
            val isLatest = currentViewRequestId.get() == thisRequestId
            if (!isLatest) {
                PoltioLogger.debug { "Ignoring outdated widget resolution result for '$targetURL' (newer screen was already requested)." }
                return@resolveMobileWidget
            }

            result.onSuccess { widget ->
                widgetCache.set(CachedWidgetResult.Widget(widget), targetURL)
                PoltioOverlayManager.showTrigger(widget, activePuid)
            }.onFailure { error ->
                PoltioOverlayManager.hideTrigger()
                if (error is PoltioNoWidgetException) {
                    widgetCache.set(CachedWidgetResult.NoWidget, targetURL)
                }
                PoltioLogger.warning { "Widget resolution skipped/failed for '$targetURL': ${error.message}" }
            }
        }

        synchronized(lock) {
            if (currentViewRequestId.get() == thisRequestId) {
                activeCall = call
            } else {
                // A newer screen already superseded this request before it could be stored;
                // cancel it explicitly instead of letting it run to completion unused.
                call.cancel()
            }
        }
    }

    // MARK: - Public Conversion Tracking API

    /**
     * Records a completed purchase for conversion attribution ("recordMobilePurchase"). Books
     * the purchase against the Poltio session already recorded for this device (via
     * `track(event = "view", ...)`), crediting revenue back to the recommendation that led to
     * it. Fire-and-forget: performs the request on a background thread, suppresses all errors,
     * and never blocks or throws — safe to call from a checkout-success handler.
     *
     * @param orderId Unique order/transaction identifier. Used by the backend as a deduplication
     * key — retrying with the same `orderId` records the purchase once. Reusing an `orderId`
     * across distinct purchases silently drops revenue, so always pass a fresh one per order.
     * @param value Total monetary value of the purchase. Must be greater than zero.
     * @param url The checkout/success screen URL or deep link (e.g. "myapp://checkout/complete").
     * Must include a scheme and host.
     * @param currency Optional ISO 4217 currency code (e.g. "USD").
     * @param items Optional line items included in the purchase.
     * @param eventTimeSeconds Optional unix timestamp (seconds) the purchase actually occurred.
     */
    fun recordPurchase(
        orderId: String,
        value: Double,
        url: String,
        currency: String? = null,
        items: List<PoltioPurchaseItem> = emptyList(),
        eventTimeSeconds: Long? = null,
    ) {
        val trimmedOrderId = orderId.trim()
        if (trimmedOrderId.isEmpty()) {
            PoltioLogger.error { "recordPurchase requires a non-empty orderId." }
            return
        }
        if (value <= 0 || !value.isFinite()) {
            PoltioLogger.error { "recordPurchase requires a positive, finite value (received $value)." }
            return
        }
        val trimmedURL = url.trim()
        if (!isValidConversionURL(trimmedURL)) {
            PoltioLogger.error { "recordPurchase requires a valid url with a scheme and host (e.g. 'myapp://checkout/complete'); received '$url'." }
            return
        }

        val key = clientKey
        if (!isInitialized || key == null) {
            PoltioLogger.warning { "recordPurchase() called before configuration. Call PoltioSDK.configure(context, clientKey) first." }
            return
        }

        val validItems = items.filter { item ->
            val hasValidId = item.id.trim().isNotEmpty()
            val hasValidValue = item.value == null || (item.value.isFinite() && item.value >= 0)
            val hasValidQuantity = item.quantity == null || item.quantity > 0
            val isValid = hasValidId && hasValidValue && hasValidQuantity
            if (!isValid) {
                PoltioLogger.warning { "recordPurchase ignoring invalid item '${item.id}' for order '$trimmedOrderId'." }
            }
            isValid
        }

        apiClient.recordPurchase(
            clientKey = key,
            deviceId = sdkId,
            url = trimmedURL,
            orderId = trimmedOrderId,
            value = value,
            currency = currency,
            eventTimeSeconds = eventTimeSeconds,
            items = validItems,
        )
    }

    // MARK: - Internal Impression Reporting

    /**
     * Reports a widget impression ("cta-view") to the backend. Called exactly once by
     * [PoltioOverlayManager] each time a floating trigger is actually displayed on screen
     * (including when the widget came from cache) — never on resolution alone, and never for a
     * trigger that's suppressed (`floating-hide-button`, within its close-remember window, or an
     * unsupported type). Fire-and-forget and safe to call before configuration; never blocks or throws.
     */
    internal fun reportCtaView(widget: PoltioWidgetResponse) {
        val key = clientKey
        if (!isInitialized || key == null) {
            PoltioLogger.debug { "reportCtaView skipped — SDK not configured." }
            return
        }
        apiClient.reportCtaView(key, sdkId, widget.publicId, widget.widgetId)
    }

    // MARK: - Public Widget Event Bridge

    /**
     * Optional callback invoked when the widget WebView emits a bridge event (e.g. "close",
     * "complete", "leadSubmit"). Delivered on the main thread. Set this once, e.g. alongside
     * `configure()`.
     */
    var onWidgetEvent: ((event: String, data: Map<String, Any?>?) -> Unit)? = null

    // MARK: - Internal Helpers

    /** Checks whether an event name corresponds to a view event. */
    internal fun isViewEvent(eventName: String): Boolean {
        val lower = eventName.lowercase()
        return lower == "view" || lower == "viewcontent" || lower == "view_content"
    }
}
